File: l23.py
"""Merge k Sorted Lists"""

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        dummyhead = ListNode(0)
        curr = dummyhead
        while None in lists:
            lists.remove(None)
        tmp = []
        for i in range(len(lists)):
            tmp.append(lists[i].val)

        while lists:                     
            a = tmp.index(min(tmp))
            curr.next = lists[a] 
            curr = curr.next
            if lists[a].next:
                lists[a] = lists[a].next
                tmp[a] = lists[a].val
            else:
                lists.remove(lists[a])
                tmp.remove(tmp[a])

        return dummyhead.next
            
               
a = Solution()
print (a.mergeKLists([[]]))

# beat 6% people ...

"""
    this is the answer from other

    from Queue import PriorityQueue 

        dummy = ListNode(None)
        curr = dummy
        q = PriorityQueue()
        for node in lists:
            if node: q.put((node.val,node))
        while q.qsize()>0:
            curr.next = q.get()[1]
            curr=curr.next
            if curr.next: q.put((curr.next.val, curr.next))
        return dummy.next   

"""

        

File: l23.java
import java.util.Comparator;
import java.util.PriorityQueue;

// Merge k Sorted Lists

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        Comparator<ListNode> cmp = (n1, n2) -> Integer.compare(n1.val, n2.val);
        PriorityQueue<ListNode> queue = new PriorityQueue<>(cmp);
        ListNode dummy = new ListNode(0);
        ListNode cur = dummy;

        for (int i=0; i < lists.length; i++) {
            if (lists[i] != null) {
                queue.offer(lists[i]);
            }
        }

        while (!queue.isEmpty()) {
            ListNode node = queue.poll();
            cur.next = node;
            cur = cur.next;
            if (node.next != null) {
                queue.offer(node.next);
            }
        }
        return dummy.next;
    }
}